### PR TITLE
fix(connector): [Adyen] `ErrorHandling` in case of Balance Check for Gift Cards

### PR DESCRIPTION
**Pull request by Sakilmostak**
`##` Type of Change

- [x] Bugfix

## Description

5xx error was being thrown in case of low balance which was unable to show the error to the merchant. Fix include failing the payment along with the error message in case of low balance

## How did you test it?
Test cases added:
Success Case: 
`{
    "amount": 4100,
    "currency": "EUR",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 100,
    "customer_id": "StripeCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "gift_card",
    "payment_method_type": "givex",
    "payment_method_data": {
        "gift_card": {
            "givex": {
                "number": "6364530000000000",
                "cvc": "122222"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "GB",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "GB",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}`
Failure Case: 
`{
    "amount": 14100,
    "currency": "EUR",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 100,
    "customer_id": "StripeCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "gift_card",
    "payment_method_type": "givex",
    "payment_method_data": {
        "gift_card": {
            "givex": {
                "number": "6364530000000000",
                "cvc": "122222"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "GB",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "GB",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}`
Screenshot 2023-10-05 at 7 29 41 PM
Screenshot 2023-10-05 at 7 29 48 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by ArjunKarthik** [comment]
Please fix the conflicts and add test screenshots

File: crates/router/src/connector/adyen/transformers.rs [hunk 1]
```diff
@@ -397,8 +397,8 @@ pub enum ActionType {
 
 #[derive(Default, Debug, Clone, Serialize, Deserialize)]
 pub struct Amount {
-    currency: String,
-    value: i64,
+    pub currency: String,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:400]: Let's have enum for currency

File: crates/router/src/connector/adyen.rs [hunk 2]
```diff
@@ -758,37 +705,67 @@ impl
...
+                field_name: "amount",
+            })?,
+        };
+
+        if response.balance.currency != currency || response.balance.value < amount {
+            Ok(types::RouterData {
+                response: Err(types::ErrorResponse {
+                    code: consts::NO_ERROR_CODE.to_string(),
+                    message: consts::LOW_BALANCE_ERROR_MESSAGE.to_string(),
+                    reason: None,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen.rs:756]: pass LOW_BALANCE_ERROR_MESSAGE in reason also

File: crates/router/src/connector/adyen.rs [hunk 3]
```diff
@@ -758,37 +705,67 @@ impl
...
+            None => Err(errors::ConnectorError::MissingRequiredField {
+                field_name: "amount",
+            })?,
+        };
+
+        if response.balance.currency != currency || response.balance.value < amount {
+            Ok(types::RouterData {
+                response: Err(types::ErrorResponse {
+                    code: consts::NO_ERROR_CODE.to_string(),
+                    message: consts::LOW_BALANCE_ERROR_MESSAGE.to_string(),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen.rs:755]: for message: pass NO_ERROR_MESSAGE

File: crates/router/src/consts.rs [hunk 4]
```diff
@@ -21,6 +21,7 @@ pub const DEFAULT_FULFILLMENT_TIME: i64 = 15 * 60;
 pub(crate) const NO_ERROR_MESSAGE: &str = "No error message";
 pub(crate) const NO_ERROR_CODE: &str = "No error code";
 pub(crate) const UNSUPPORTED_ERROR_MESSAGE: &str = "Unsupported response type";
+pub(crate) const LOW_BALANCE_ERROR_MESSAGE: &str = "Balance in the payment method is low";
```
Comment (ArjunKarthik) [inline crates/router/src/consts.rs:24]: Insufficient balance in the payment method

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — APPROVED** [review]

File: crates/router/src/connector/adyen/transformers.rs [hunk 5]
```diff
@@ -2867,7 +2879,7 @@ impl TryFrom<types::PaymentsBalanceResponseRouterData<AdyenBalanceResponse>>
             }),
             payment_method_balance: Some(types::PaymentMethodBalance {
                 amount: item.response.balance.value,
-                currency: item.response.balance.currency,
+                currency: item.response.balance.currency.to_string(),
```
Comment (srujanchikke) [inline crates/router/src/connector/adyen/transformers.rs:2882]: Can this be also changed to `storage_enums::Currency`

**Review by srujanchikke — APPROVED** [review]

File: crates/router/src/connector/adyen.rs [hunk 6]
```diff
@@ -803,19 +752,47 @@ impl
...
+        };
+
+        if response.balance.currency != currency || response.balance.value < amount {
+            Ok(types::RouterData {
+                response: Err(types::ErrorResponse {
+                    code: consts::NO_ERROR_CODE.to_string(),
+                    message: consts::NO_ERROR_MESSAGE.to_string(),
+                    reason: Some(consts::LOW_BALANCE_ERROR_MESSAGE.to_string()),
+                    status_code: res.status_code,
+                    attempt_status: None,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen.rs:783]: send failed status

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**